src/NServiceBus.Persistence.DynamoDB/Serialization/MemoryStreamConverter.cs

```diff
@@ -61,14 +65,33 @@ public static bool TryExtract(JsonProperty property, out MemoryStream? memoryStr
             {
                 return false;
             }
-            memoryStream = new MemoryStream(property.Value.GetBytesFromBase64());
+
+            GetStream(property.Value.GetGuid(), out memoryStream);
             return true;
         }
 
-        public static JsonNode ToNode(MemoryStream memoryStream) =>
-            new JsonObject
+        public static JsonNode ToNode(MemoryStream memoryStream)
+            => new JsonObject
             {
-                [PropertyName] = Convert.ToBase64String(memoryStream.ToArray())
+                [PropertyName] = TrackStream(memoryStream)
             };
+
+        static void GetStream(Guid streamId, out MemoryStream? memoryStream)
+        {
+            if (StreamMap.Value!.TryGetValue(streamId, out memoryStream))
+            {
+                StreamMap.Value.Remove(streamId);
+            }
+        }
+
+        static Guid TrackStream(MemoryStream memoryStream)
+        {
+            var streamId = Guid.NewGuid();
+            StreamMap.Value!.Add(streamId, memoryStream);
+            return streamId;
+        }
+
+        // internal for tests
+        internal static readonly ThreadLocal<Dictionary<Guid, MemoryStream>> StreamMap = new(() => new Dictionary<Guid, MemoryStream>());
```

Comment:
I think the dictionary makes sense but strictly speaking, I'm not sure I can see a valid code path that would involve multiple calls on the same thread at the same time since none of the involved APIs is async?

Reply:
A class with multiple properties that are of type `MemoryStream` or a class that has `HashSet<MemoryStream>`